crates/task-impls/src/quorum_proposal.rs

```diff
@@ -0,0 +1,574 @@
+            )),
+            &self.sender,
+        )
+        .await;
+        broadcast_event(
+            Arc::new(HotShotEvent::DummyQuorumProposalSend(self.view_number)),
+            &self.sender,
+        )
+        .await;
+    }
+}
+
+/// The state for the quorum proposal task.
+pub struct QuorumProposalTaskState<TYPES: NodeType, I: NodeImplementation<TYPES>> {
+    /// Latest view number that has been proposed for.
+    pub latest_proposed_view: TYPES::Time,
+
+    /// Table for the in-progress proposal depdencey tasks.
+    pub propose_dependencies: HashMap<TYPES::Time, JoinHandle<()>>,
+
+    /// Network for all nodes
+    pub quorum_network: Arc<I::QuorumNetwork>,
+
+    /// Network for DA committee
+    pub committee_network: Arc<I::CommitteeNetwork>,
+
+    /// Output events to application
+    pub output_event_stream: async_broadcast::Sender<Event<TYPES>>,
+
+    /// Reference to consensus. The replica will require a write lock on this.
+    pub consensus: Arc<RwLock<Consensus<TYPES>>>,
+
+    /// Membership for Timeout votes/certs
+    pub timeout_membership: Arc<TYPES::Membership>,
+
+    /// Membership for Quorum Certs/votes
+    pub quorum_membership: Arc<TYPES::Membership>,
+
+    /// The node's id
+    pub id: u64,
+}
+
+impl<TYPES: NodeType, I: NodeImplementation<TYPES>> QuorumProposalTaskState<TYPES, I> {
+    /// Create an event dependency
+    #[instrument(skip_all, fields(id = self.id, latest_proposed_view = *self.latest_proposed_view), name = "Quorum proposal create event dependency", level = "error")]
+    fn create_event_dependency(
+        &self,
+        dependency_type: ProposalDependency,
+        view_number: TYPES::Time,
+        event_receiver: Receiver<Arc<HotShotEvent<TYPES>>>,
+    ) -> EventDependency<Arc<HotShotEvent<TYPES>>> {
+        EventDependency::new(
+            event_receiver,
+            Box::new(move |event| {
+                let event = event.as_ref();
+                let event_view = match dependency_type {
+                    ProposalDependency::ViewSync => {
+                        if let HotShotEvent::QuorumProposalViewSyncFinalizeCertValidated(
+                            view_sync_cert,
+                        ) = event
+                        {
+                            view_sync_cert.view_number
+                        } else {
+                            return false;
+                        }
+                    }
+                    ProposalDependency::QCFormed => {
+                        if let HotShotEvent::QuorumProposalTimeoutCertValidated(timeout_cert) =
+                            event
+                        {
+                            timeout_cert.view_number
+                        } else if let HotShotEvent::QuorumProposalQuorumCertValidated(qc) = event {
+                            qc.view_number
+                        } else {
+                            return false;
+                        }
+                    }
```

Comment:
We can probably merge these two enum variants into one `ProposalDependency::CertValidated` (or a better name) and use it to match the three `HotShotEvent::*CertValidated` events. They are "or" dependencies anyway.

Reply:
Great idea! Added.